src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -848,22 +848,53 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
             return sum;
         }
 
-        internal static Dictionary<K, ImmutableArray<T>> ToDictionary<K, T>(this ImmutableArray<T> items, Func<T, K> keySelector)
+        /// <summary>
+        /// Helper type to give us a PooledDictionary cache in the case where the key for the dictionary is a <see
+        /// cref="ReadOnlyMemory{T}"/>.  These dictionaries need a special comparer, so we cannot use the standard
+        /// pattern of <c>PooledDictionary.GetInstance()</c>.
+        /// </summary>
+        private static class AccumulatorCache<K>
+             where K : notnull
+        {
+            private static readonly ObjectPool<PooledDictionary<ReadOnlyMemory<char>, object>> s_readOnlyMemoryCache =
+                PooledDictionary<ReadOnlyMemory<char>, object>.CreatePool(ReadOnlyMemoryOfCharComparer.Instance);
+
+            public static PooledDictionary<K, object> GetInstance(IEqualityComparer<K>? equalityComparer)
+            {
+                if (typeof(K) == typeof(ReadOnlyMemory<char>))
```

Comment:
>if (typeof(K) == typeof(ReadOnlyMemory<char>)) [](http://example.com/codeflow?start=16&length=46)

I would prefer to not have this magic logic here

Reply:
k.

Reply:
done.